### ISSUE TITLE
Expr arithmetic op return type inconsistency when encountering null values

### ISSUE THREAD

**Issue by harahu** [pola-rs/polars#8224]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://crates.io/crates/polars) of Polars.

### Issue description

In the aggregation context, arithmetic operations sometimes return inconsistent data types. In particular, when subtracting an already aggregated value from a column, whether or not the operation aggregates or not is dependent on whether the subtracted value is always null.

```rust
// Results in scalar null
let res = col("foo").sub(scalar_always_null);

// Results in list of same size as `col("foo")`
let res = col("foo").sub(scalar_sometimes_or_always_non_null);
```

This is problematic because it means one cannot reason effectively about the resulting column type, which has consequences for downstream operations.

For instance, following up with a check like:

```rust
res.is_null().all()
```

Will either successfully return a boolean column or panic, depending on the code path.

### Reproducible example

```rust
use std::ops::Sub;

use polars::{
    lazy::dsl::{col, Expr},
    prelude::{DataFrame, IntoLazy, NamedFrom, Series},
};

fn sub_col_min(column: &str, min_column: &str) -> Expr {
    col(column).sub(col(min_column).min())
}

#[test]
fn test_nulls_broadcast_failure_repro() {
    let group = Series::new("group", [1, 1, 2, 2]);
    let vals_null = Series::new("vals_null", [None::<f64>; 4]);
    let vals_partial = Series::new(
        "vals_partial",
        [None::<f64>, None::<f64>, Some(12.0), Some(13.0)],
    );
    let vals_num = Series::new("vals_num", [10.0, 11.0, 12.0, 13.0]);
    let dataset = DataFrame::new(vec![group, vals_null, vals_partial, vals_num]).unwrap();

    let df = dataset
        .clone()
        .lazy()
        .groupby(["group"])
        .agg([
            sub_col_min("vals_num", "vals_num").alias("sub_num"),
            sub_col_min("vals_num", "vals_partial").alias("sub_partial"),
            sub_col_min("vals_num", "vals_null").alias("sub_null"),
        ])
        .collect()
        .unwrap();

    println!("{}", df);

    assert_eq!(
        df.column("sub_num").unwrap().dtype(),
        df.column("sub_partial").unwrap().dtype(),
    );

    assert_eq!(
        df.column("sub_num").unwrap().dtype(),
        df.column("sub_null").unwrap().dtype(),
    );
}

#[test]
fn test_nulls_broadcast_failure_repro_panic() {
    let group = Series::new("group", [1, 1, 2, 2]);
    let vals_null = Series::new("vals_null", [None::<f64>; 4]);
    let vals_partial = Series::new(
        "vals_partial",
        [None::<f64>, None::<f64>, Some(12.0), Some(13.0)],
    );
    let vals_num = Series::new("vals_num", [10.0, 11.0, 12.0, 13.0]);
    let dataset = DataFrame::new(vec![group, vals_null, vals_partial, vals_num]).unwrap();

    dataset
        .clone()
        .lazy()
        .groupby(["group"])
        .agg([
            sub_col_min("vals_num", "vals_num")
                .is_null()
                .all()
                .alias("sub_num"),
            sub_col_min("vals_num", "vals_partial")
                .is_null()
                .all()
                .alias("sub_partial"),
            sub_col_min("vals_num", "vals_null")
                .is_null()
                .all()
                .alias("sub_null"),
        ])
        .collect()
        .unwrap();
}
```

### Expected behavior

No auto-aggregation when encountering all null values.

### Installed versions

<details>
polars = {version = "0.28.0", features = ["lazy"]}
</details>

**Comment by ritchie46** [comment]
Thanks for the report. Could you make a small repro in python?

**Comment by harahu** [comment]
@ritchie46 

```python
import sys
from typing import NoReturn

import polars as pl


def sub_col_min(column: str, min_column: str) -> pl.Expr:
    return pl.col(column).sub(pl.col(min_column).min())


def test_1() -> NoReturn:
    df = pl.DataFrame(
        {
            "group": [1, 1, 2, 2],
            "vals_num": [10.0, 11.0, 12.0, 13.0],
            "vals_partial": [None, None, 12.0, 13.0],
            "vals_null": [None, None, None, None],
        }
    )

    q = (
        df.lazy().groupby("group")
        .agg(
            [
                sub_col_min("vals_num", "vals_num").alias("sub_num"),
                sub_col_min("vals_num", "vals_partial").alias("sub_partial"),
                sub_col_min("vals_num", "vals_null").alias("sub_null"),
            ]
        )
    )

    df = q.collect()

    print(df)

    assert len(set(df.dtypes)) == 1


def test_2() -> NoReturn:
    df = pl.DataFrame(
        {
            "group": [1, 1, 2, 2],
            "vals_num": [10.0, 11.0, 12.0, 13.0],
            "vals_partial": [None, None, 12.0, 13.0],
            "vals_null": [None, None, None, None],
        }
    )

    df.lazy().groupby("group").agg(
        [
            sub_col_min("vals_num", "vals_num").is_null().all().alias("sub_num"),
            sub_col_min("vals_num", "vals_partial").is_null().all().alias("sub_partial"),
            sub_col_min("vals_num", "vals_null").is_null().all().alias("sub_null"),
        ]
    ).collect()


if __name__ == "__main__":
    try:
        test_1()
    except Exception as e:
        print(e, file=sys.stderr)

    try:
        test_2()
    except Exception as e:
        print(e, file=sys.stderr)
```

**Comment by ritchie46** [comment]
Thanks! Totally agree we should not have this behavior. Fix coming up.

**Comment by ritchie46** [comment]
fixed by #8322

**Comment by ritchie46** [comment]
This wasn't fast :wink:
